lib/Dialect/ESI/cosim/cosim_dpi_server/DpiEntryPoints.cpp

```diff
@@ -0,0 +1,208 @@
+           "(total_bytes==0)\n");
+    err++;
+  }
+  int numElems = svSize(data, 1);
+  int elemSize = numElems == 0 ? 0 : (totalBytes / numElems);
+  if (numElems * expectedElemSize != totalBytes) {
+    printf("DPI-C: ERROR: passed array argument that doesn't have expected "
+           "element-size: expected=%d actual=%d\n",
+           expectedElemSize, elemSize);
+    err++;
+  }
+  return err;
+}
+
+// ---- DPI entry points ----
+
+// Register simulated device endpoints.
+// - return 0 on success, non-zero on failure (duplicate EP registered).
+DPI int sv2cCosimserverEpRegister(int endpointId, long long sendTypeId,
+                                  int sendTypeSize, long long recvTypeId,
+                                  int recvTypeSize) {
+  // Ensure the server has been constructed.
+  sv2cCosimserverInit();
+  try {
+    // Then register with it.
+    server->endpoints.registerEndpoint(endpointId, sendTypeId, sendTypeSize,
+                                       recvTypeId, recvTypeSize);
+    return 0;
+  } catch (const std::runtime_error &rt) {
+    fprintf(stderr, "%s\n", rt.what());
+    return 1;
+  }
+}
+
+// Attempt to recieve data from a client.
+//   - Returns negative when call failed (e.g. EP not registered).
+//   - If no message, return 0 with size_bytes == 0.
+//   - Assumes buffer is large enough to contain entire message. Fails if not
+//     large enough. (In the future, will add support for getting the message
+//     into a fixed-size buffer over multiple calls.)
+DPI int sv2cCosimserverEpTryGet(unsigned int endpointId,
+                                // NOLINTNEXTLINE(misc-misplaced-const)
+                                const svOpenArrayHandle data,
+                                unsigned int *dataSize) {
+  if (server == nullptr)
+    return -3;
+
+  try {
+    Endpoint::BlobPtr msg;
+    // Poll for a message.
+    if (server->endpoints[endpointId].getMessageToSim(msg)) {
+
+      // Do the validation only if there's a message available. Since the
+      // simulator is going to poll up to every tick and there's not going to be
+      // a message most of the time, this is important for performance.
+
+      if (validateSvOpenArray(data, sizeof(int8_t)) != 0) {
+        printf("ERROR: DPI-func=%s line=%d event=invalid-sv-array\n", __func__,
+               __LINE__);
+        return -1;
+      }
+
+      // Detect or verify size of buffer.
+      if (*dataSize == ~0u) {
+        *dataSize = svSizeOfArray(data);
+      } else if (*dataSize > (unsigned)svSizeOfArray(data)) {
+        printf("ERROR: DPI-func=%s line %d event=invalid-size (max %d)\n",
+               __func__, __LINE__, (unsigned)svSizeOfArray(data));
+        return -2;
+      }
+      // Verify it'll fit.
+      if (msg->size() > *dataSize) {
+        printf("ERROR: Message size too big to fit in RTL buffer\n");
+        return -4;
+      }
+
+      // Copy the message data.
+      size_t i;
+      for (i = 0; i < msg->size(); i++) {
```

Comment:
not sure if it's a big deal here, but could this use pre-increment? the LLVM style [guide](https://llvm.org/docs/CodingStandards.html#prefer-preincrement) makes it seem like it can matter for performance, which other parts of the PR seem to care a lot about.

Reply:
Yep. That's something I forgot. I'm trying to match the LLVM style.